File: src/main/java/org/goafabric/dbagentnew/dbagent/logic/PersonLogic.java
package org.goafabric.dbagentnew.dbagent.logic;

import dev.langchain4j.agent.tool.Tool;
import org.goafabric.dbagentnew.dbagent.persistence.PersonRepository;
import org.goafabric.dbagentnew.dbagent.persistence.entity.Person;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.stream.StreamSupport;

@Component
//@Transactional
public class PersonLogic {

    private final PersonRepository personRepository;


    public PersonLogic(PersonRepository personRepository) {
        this.personRepository = personRepository;
    }

    public List<Person> findAll() {
        return StreamSupport.stream(personRepository.findAll().spliterator(), false)
                .toList();
    }


    @Tool
    public List<Person> findByFirstName(String firstName) {
        return personRepository.findByFirstNameIgnoreCase(firstName);
    }

    @Tool
    public List<Person> findByLastName(String lastName) {
        return personRepository.findByLastNameIgnoreCase(lastName);
    }

    @Tool
    public List<Person> findByCity(String city) {
        return personRepository.findByAddressCityContainsIgnoreCase(city);
    }

    @Tool
    public List<Person> findByAllergy(String allergy) {
        return personRepository.findByAllergyAllergyContainsIgnoreCase(allergy);
    }

    public Person save(Person person) {
        return personRepository.save(person);
    }

}

File: src/main/kotlin/org/goafabric/llm/dbagent/logic/PersonLogic.kt
package org.goafabric.llm.dbagent.logic

import dev.langchain4j.agent.tool.Tool
import org.goafabric.llm.dbagent.persistence.PersonRepository
import org.goafabric.llm.dbagent.persistence.entity.Person
import org.springframework.stereotype.Component
import java.util.stream.StreamSupport

@Component //@Transactional
class PersonLogic(private val personRepository: PersonRepository) {
    fun findAll(): MutableList<Person> {
        return StreamSupport.stream<Person>(personRepository.findAll().spliterator(), false)
            .toList()
    }


    @Tool
    fun findByFirstName(firstName: String): MutableList<Person> {
        return personRepository.findByFirstNameIgnoreCase(firstName)
    }

    @Tool
    fun findByLastName(lastName: String): MutableList<Person> {
        return personRepository.findByLastNameIgnoreCase(lastName)
    }

    @Tool
    fun findByCity(city: String): MutableList<Person> {
        return personRepository.findByAddressCityContainsIgnoreCase(city)
    }

    @Tool
    fun findByAllergy(allergy: String): MutableList<Person> {
        return personRepository.findByAllergyAllergyContainsIgnoreCase(allergy)
    }

    fun save(person: Person): Person {
        return personRepository.save<Person>(person)
    }
}
